yew/src/virtual_dom/vlist.rs

```diff
@@ -94,84 +88,590 @@ impl VDiff for VList {
+            // at the end of the algorithm, but we will steal nodes from
+            // `rights` so will not be able to use it.
+            let right_keys: Vec<Key> = rights
+                .iter()
+                .map(|vnode| vnode.key().clone().expect("right must have a key"))
+                .collect();
+
+            // We will try to match the left vnodes with the right ones, and
+            // store them in `matched_rights`: it is formed of vnodes from
+            // right, in the same order than left. `reused_rights` is used to
+            // know which right nodes have been matched and reused.
+            let mut matched_rights: Vec<Option<VNode>> = vec![None; self.children.len()];
+            let mut reused_rights = FixedBitSet::with_capacity(rights.len());
+
+            // Generator of dummy vnodes, used when stealing vnodes from rights.
+            let make_dummy = || {
+                VNode::VText(VText {
+                    text: String::default(),
+                    reference: None,
+                })
+            };
+
+            // Match all the elements that are not new, but may have moved.
+            let mut new_lefts: HashSet<usize> = HashSet::with_capacity(self.children.len());
+            for ((left_idx, left), matched_right) in self
+                .children
+                .iter()
+                .enumerate()
+                .zip(matched_rights.iter_mut())
+            {
+                let left_key = left.key().as_ref().expect("must have a key");
+                if let Some(right_idx) = rights_lookup.remove(&left_key) {
+                    let right = rights
+                        .get_mut(right_idx)
+                        .expect("the index from the map must be valid");
+                    let right = core::mem::replace(right, make_dummy());
+                    *matched_right = Some(right);
+                    reused_rights.put(right_idx);
+                } else {
+                    new_lefts.insert(left_idx);
+                }
+            }
+            let new_lefts = new_lefts; // remove mutability
+
+            // Reconciliation loop, i.e. apply the least amount of
+            // transformations to rights to make them identical to lefts.
+            let mut matched_rights = matched_rights.into_iter().peekable();
+            for left in self.children.iter_mut() {
+                let ancestor = matched_rights.next().unwrap();
+
+                // Create a new `next_sibling` reference which points to the
+                // next `right` or the outer list's `next_sibling` if there
+                // are no more `rights`.
                 let new_next_sibling = NodeRef::default();
-                if let Some(next_right) = rights.peek() {
+                if let Some(Some(next_right)) = matched_rights.peek() {
                     new_next_sibling.set(Some(next_right.first_node()));
                 } else {
                     new_next_sibling.link(next_sibling.clone());
```

Comment:
@totorigolo  I think this comment got lost: https://github.com/yewstack/yew/pull/1231#discussion_r438823306

> If the next right is unmatched (but is not the last one), we enter this branch and could place a mid-list left at the end of the list. I think we need to move rights before the reconciliation loop to avoid this issue.

This affects keyed items that have different node type (have to be recreated)